### ISSUE TITLE
Very different layout in SamsungGalaxyS7 skin and on real S7 device

### ISSUE THREAD

**Issue by ThomasH99** [codenameone/CodenameOne#3741]
**Describe the bug**
The layout (and resolution) on a Samsung S7 device and in the simulator S7 skin are very different, making it difficult to fine-tune the layout with the simulator. See the attached screenshots. Basically my app seems to start up in 1920x1080 even though the device is set to 2560x1440. 

Also, the space between text lines in the TextArea (with random text) is too large and it looks awful in the medium resolution. However, even when forcing the device to the high resolution, the text is not shown in the same way as in the simulator - it takes 4 lines in the simulator and 3 on the device (although the fonts should be the same?). 

I've noticed that after starting my app, I can get the full resolution by tricking the app: I set the device to medium resolution (1920x1080), start the app, then go directly to the Samsung device settings and change resolution to 2560x1440 and then switch back to the app again. This way I can see the full resolution. 

At one point while experimenting I got an error message on the Samsung: 

"An internal application error occurred:
java.lang.OutOfMemoryError:
Failed to allocate a 102972144
byte allocation with 25165824
free bytes and 91MB until 00M, max allowed footprint
130904280, growth limit 201326592"

This makes me wonder if the device itself (or CN1) switches down the resolution if there is a lack of memory?

Here are the screenshots:

Samsung S7 skin (Simulator)

![CodenameOne Screenshot 15](https://github.com/codenameone/CodenameOne/assets/16265939/0baf525a-7aa3-43ba-976d-a551a31d9df1)

Samsung S7 actual screenshot (too low resolution 1920x1080)

![Screenshot_20230930-225737](https://github.com/codenameone/CodenameOne/assets/16265939/01b30b22-bd4d-4e5e-9278-6ea50cded87d)

For reference, a device screenshot after forcing the resolution to 2560x1440 as described above: 

![Screenshot_20230930-225713](https://github.com/codenameone/CodenameOne/assets/16265939/de4db9b6-0f92-4b35-8ad0-96166d247386)

**Expected behavior**
The resolution and visual appearance should be the same on the simulator and the device. 

**Smartphone (please complete the following information):**
 - Device: Samsung Galaxy S7 (4Gb memory, 1.8Gb free). 
 - OS: Android 8.0.0. Samsung Experience 9.0

**Comment by shai-almog** [comment]
Cam you create a simulator shot without the device skin so all images would be 1440 × 2560 in size? 

That way I can measure the distances accurately.

**Comment by ThomasH99** [comment]
Here is the simulator screenshot without the device skin. I noticed that whether you create it 'normally' or zoomed you get the full resolution which is neat :-)
![CodenameOne Screenshot 18](https://github.com/codenameone/CodenameOne/assets/16265939/b0a05702-a824-4015-b49f-c444b6560cdb)

**Comment by shai-almog** [comment]
Thanks.
I'm looking over these and while the simulator screenshot is 1440 × 2560 as it should be. The device screenshot is 1080 × 1920 and is a JPG instead of a PNG. 

How did you grab the screenshot? 

Did you use the power and volume button together?

How did you transfer the file? If you used an app like whatsapp, they often scale the image file.

**Comment by ThomasH99** [comment]
Power+sound doesn’t work on S7, you need to use a (ridiculously impractical) ‘palm sweep’. But it is the native screenshot feature. I transferred the screenshots by emailing via gmail (no ‘reduce picture resolution?’ questions asked), then drag&drop on mac into the github issue. Maybe the device reduces the screenshot to the ‘forced’ medium (1920x1080) resolution?The high resolution screenshot (that I tricked the app into) was it also only medium resolution? Because that would seem to confirm a reduction took place somewhere. Sent from my iPhoneOn 2 Oct 2023, at 07:43, Shai Almog ***@***.***> wrote:﻿
Thanks.
I'm looking over these and while the simulator screenshot is 1440 × 2560 as it should be. The device screenshot is 1080 × 1920 and is a JPG instead of a PNG.
How did you grab the screenshot?
Did you use the power and volume button together?
How did you transfer the file? If you used an app like whatsapp, they often scale the image file.

**Comment by ThomasH99** [comment]
I checked again, Galaxy S7 does not support PNG (only came in Samsung versions after 8.0 (but screenshots CAN done by pushing buttons Power+Select for 2-5 s). 


So, what I sent is indeed the original unmodified screenshots. But I found one which is in full res 1440 × 2560 while showing the text in ‘medium' resolution. I hope it’ll do.


￼

**Comment by shai-almog** [comment]
Great! But it seems to be missing from the issue? Can you attach it?

**Comment by ThomasH99** [comment]
Galaxy S7 does not support PNG (only came in Samsung versions after 8.0. 
Here's a screenshot which is in full res 1440 × 2560 while showing the text in ‘medium' resolution. 
![Screenshot_20230930-225455](https://github.com/codenameone/CodenameOne/assets/16265939/5f7b8d0f-b3c2-4041-b590-979d69c434c4)

**Comment by shai-almog** [comment]
This isn't a full resolution image. Did something break in the attachment?

**Comment by ThomasH99** [comment]
Sorry, not sure what went wrong. Let me try again. To test this time, I downloaded the picture from the issue Preview tab and checked that the resolution is 2560x1440 (but alas still only jpeg). Hope it gets across correctly now. 

![Screenshot_20230930-225455](https://github.com/codenameone/CodenameOne/assets/16265939/d9f0b88b-ef37-4eb6-aa63-c0b6f8bfbaa9)

**Comment by shai-almog** [comment]
Try using the skin file from here: https://github.com/codenameone/codenameone-skins/actions/runs/6408557074

**Comment by ThomasH99** [comment]
Hi, 

It doesn’t seem to be working. 

I downloaded the file (called “6408557074”) and renamed it to SamsungGalaxyS7.skin.

When I load it i the simulator (Skins - Add New...), I get the below error trace and the app later crashes.

Adding CEF to classpath
Retina Scale: 2.0
Failed loading the skin file: file:/Users/thomashjelm/NetBeansProjects/todocatalyst/New%20skin%20for%20Samsung%20Galaxy%20S7/SamsungGalaxyS7.skin
java.lang.NullPointerException
	at com.codename1.impl.javase.JavaSEPort.initializeCoordinates(JavaSEPort.java:2421)
	at com.codename1.impl.javase.JavaSEPort.loadSkinFile(JavaSEPort.java:2642)
	at com.codename1.impl.javase.JavaSEPort.loadSkinFile(JavaSEPort.java:5240)
	at com.codename1.impl.javase.JavaSEPort.init(JavaSEPort.java:5571)
	at com.codename1.impl.javase.cef.JavaCEFSEPort.init(JavaCEFSEPort.java:67)
	at com.codename1.impl.CodenameOneImplementation.initImpl(CodenameOneImplementation.java:178)
	at com.codename1.ui.Display.init(Display.java:531)
	at com.codename1.impl.javase.Executor$3.run(Executor.java:274)
	at java.desktop/java.awt.event.InvocationEvent.dispatch$$$capture(InvocationEvent.java:313)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

**Comment by ThomasH99** [comment]
Come to think about it, the issue was not the skin in the simulator, but the appearance on the actual device (which starts up with too low resolution)?

**Comment by shai-almog** [comment]
Click the link to go to the page. There scroll to the bottom where you can see a link to the skin file. Download it. Unzip the download. That should be the skin file.

**Comment by ThomasH99** [comment]
As I wrote, the issue was that the resolution was too low *on the device* - the skin won’t change that. Could you review the mail thread again? It looked like maybe it could be a memory management problem. There was also the issue of too high gap between text lines when wrapping long texts around. Basically, I can’t get a decent look on my S7. I would have liked to test on another Android device but unfortunately I don’t currently have access to any. Sent from my iPhoneOn 5 Oct 2023, at 04:01, Shai Almog ***@***.***> wrote:﻿
Click the link to go to the page. There scroll to the bottom where you can see a link to the skin file. Download it. Unzip the download. That should be the skin file.

**Comment by shai-almog** [comment]
We need a baseline. If the simulator is inaccurate (which it was) I can't measure why the difference occurs. Android specifically is very problematic with PPI and Samsung doubly so because of their gaming optimization nonsense. 

The simulator should align better with the device now. If it doesn't that might be a bug. If you think the device has the wrong choice for PPI that might be a separate bug. Unfortunately, that's a very hard bug to determine.

**Comment by ThomasH99** [comment]
Yes, I understand. I would have liked to test on another (and more recent) Android device, but iOS is my first priority, so I’m simply experimenting with what I already have. Initially, I thought it would be simple, but we can leave it for now. I’ll come back to it when it becomes a real priority. Sent from my iPadOn 6 Oct 2023, at 04:29, Shai Almog ***@***.***> wrote:﻿
We need a baseline. If the simulator is inaccurate (which it was) I can't measure why the difference occurs. Android specifically is very problematic with PPI and Samsung doubly so because of their gaming optimization nonsense.
The simulator should align better with the device now. If it doesn't that might be a bug. If you think the device has the wrong choice for PPI that might be a separate bug. Unfortunately, that's a very hard bug to determine.
